src/lib.rs

```diff
@@ -113,51 +113,85 @@ where
     type Item = SdrRecord;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let current_id = self.next_id?;
+        let mut recoverable_errors = 0;
+        const MAX_RECOVERABLE_ERRORS: usize = 10; // Max number of recoverable errors
 
-        if current_id.is_last() {
-            self.next_id.take();
-            return None;
-        }
-
-        let next_record = self
-            .ipmi
-            .send_recv(sdr::GetDeviceSdr::new(None, current_id));
-
-        let (value, next_record_id) = match next_record {
-            Ok(record) => {
-                let next_record_id = record.next_entry;
-
-                (Some(record.record), next_record_id)
-            }
-            Err(IpmiError::ParsingFailed {
-                error: ParseResponseError::Parse((e, next_record_id)),
-                ..
-            }) => {
-                log::warn!(
-                    "Recoverable error while parsing SDR record 0x{:04X}: {e:?}",
-                    current_id.value()
-                );
-                (None, next_record_id)
-            }
-            Err(e) => {
-                log::error!(
-                    "Unrecoverable error while parsing SDR record 0x{:04X}: {e:?}",
-                    current_id.value()
-                );
+        while let Some(current_id) = self.next_id {
+            if current_id.is_last() {
                 self.next_id.take();
+                if recoverable_errors > 0 {
+                    log::warn!(
+                        "Iteration completed with {} recoverable errors.",
+                        recoverable_errors
+                    );
+                }
                 return None;
             }
-        };
 
-        if next_record_id == current_id {
-            log::error!("Got duplicate SDR record IDs! Stopping iteration.");
-            self.next_id.take();
-            return None;
-        } else {
-            self.next_id = Some(next_record_id);
+            let next_record = self
+                .ipmi
+                .send_recv(sdr::GetDeviceSdr::new(None, current_id));
+
+            match next_record {
+                Ok(record) => {
+                    if recoverable_errors > 0 {
+                        log::warn!(
```

Comment:
This message feels duplicate (since the message on line 170 will always "precede" it. Would prefer to remove this one and bump the one on line 170 to `warn`.